tapir/shifts/services/slot_modification_service.py

```diff
@@ -0,0 +1,194 @@
+
+    @classmethod
+    def pick_slot_template_from_shift(
+        cls, parameter_set: ParameterSet, shift_template: ShiftTemplate
+    ) -> ShiftSlotTemplate:
+        # The shift is assumed to have been picked according to the parameters, here we only look at the slot
+
+        candidate_slot_templates = shift_template.slot_templates.filter(
+            name=parameter_set.origin_slot_name
+        )
+
+        if not candidate_slot_templates.exists():
+            raise ShiftSlotTemplate.DoesNotExist(
+                f"Could not find slot template with name {parameter_set.origin_slot_name} "
+                f"on shift template {parameter_set.origin_slot_name}"
+            )
+
+        if candidate_slot_templates.count() == 1:
+            return candidate_slot_templates.first()
+
+        # if possible, pick a slot that doesn't have an attendance
+        existing_attendance_templates = ShiftAttendanceTemplate.objects.filter(
+            slot_template__in=candidate_slot_templates
+        )
+        slot_templates_without_attendance = candidate_slot_templates.exclude(
+            attendance_template__in=existing_attendance_templates
+        )
+        if slot_templates_without_attendance.count() == 1:
+            return slot_templates_without_attendance.first()
+        if slot_templates_without_attendance.count() > 1:
+            candidate_slot_templates = slot_templates_without_attendance
+
+        return candidate_slot_templates.order_by("-id").first()
+
+    @classmethod
+    def get_affected_members(cls, changes: dict[ParameterSet, list[ShiftSlotTemplate]]):
+        result = io.StringIO()
+        writer = csv.writer(result)
+        writer.writerow(
+            ["member_id", "member_email", "member_name", "change", "slot", "warnings"]
+        )
+
+        for parameter_set, slot_templates in changes.items():
+            for slot_template in slot_templates:
+                already_written_user = None
+                if hasattr(slot_template, "attendance_template"):
+                    tapir_user = slot_template.attendance_template.user
+                    already_written_user = tapir_user
+                    cls.write_csv_row(writer, tapir_user, slot_template, parameter_set)
+                affected_attendances = ShiftAttendance.objects.filter(
+                    slot__in=slot_template.generated_slots.all(),
+                    slot__shift__start_time__gt=timezone.now(),
+                ).with_valid_state()
+                if already_written_user:
+                    affected_attendances = affected_attendances.exclude(
+                        user=already_written_user
+                    )
+                for attendance in affected_attendances:
+                    cls.write_csv_row(
+                        writer, attendance.user, attendance.slot, parameter_set
+                    )
+
+        return result.getvalue()
+
+    @classmethod
+    def write_csv_row(
```

Comment:
Trying to understand the idea, what'll happen to this CSV file?

Reply:
It's printed when calling `preview_changes` or `apply_changes`. It's the list of members that lost a slot in the process, so that we can contact them and explain why they lost their spot.

Reply:
I see. Why to build a csv, when we need to print it only?

Reply:
I find it's shorter and more reliable that trying to do shift concatenation. 
The output is also fairly clear and easy to work on by other people if needed.

Reply:
Right, good point.